src/widget/woverview.cpp

```diff
@@ -556,82 +597,137 @@ void WOverview::drawMarks(QPainter* pPainter, const float offset, const float ga
     shadowFont.setWeight(99);
     shadowFont.setPixelSize(10 * m_scaleFactor);
 
-    for (const auto& currentMark : m_marks) {
-        const WaveformMarkProperties& markProperties = currentMark->getProperties();
-        if (currentMark->isValid() && currentMark->getSamplePosition() >= 0.0) {
-            // Marks are visible by default.
-            if (currentMark->hasVisible() && !currentMark->isVisible()) {
-                continue;
-            }
+    // Text labels are rendered so they do not overlap with other WaveformMark's
+    // labels. If the text would be too wide, it is elided. However, the user
+    // can hover the mouse cursor over a label to show the whole label text,
+    // temporarily hiding any following labels that would be drawn over it.
+    // This requires looping over the WaveformMarks twice and the marks must be
+    // sorted in the order they appear on the waveform.
+    // In the first loop, the lines are drawn and the text to render plus its
+    // location are calculated. The text must be drawn in the second loop to
+    // prevent the lines of following WaveformMarks getting drawn over it.
+    QList<WaveformMarkPointer> marksToRender;
+    for (WaveformMarkPointer mark : m_marks) {
+        if (mark->isValid() && mark->getSamplePosition() >= 0.0 && mark->isVisible()) {
+            marksToRender.append(mark);
+        }
+    }
+    std::sort(marksToRender.begin(), marksToRender.end());
```

Comment:
I guess the result of the block above is almost constant. Can we move it out if the painting function?

Reply:
Yes, good point